File: packages/sandboxd/src/pi_proxy/state.rs
//! Shared mutable state for the Pi proxy runtime.
//!
//! The state serializes commands to the stdio RPC child, fans out Pi events to
//! websocket clients, and projects agent activity into sandbox keepalive.

use std::io::Write;
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::mpsc::{self, Receiver, RecvTimeoutError, Sender};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, Instant};

use serde_json::{Value, json};

use crate::cgroups::{attach_pid_to_scope, kill_scope};
use crate::keepalive::KeepaliveManager;
use crate::pi_proxy::idempotency::SharedIdempotencyStore;
use crate::pi_proxy::json_rpc::render_pi_event_json_rpc_notification;
use crate::pi_proxy::rpc_process::{PiRpcChild, PiRpcOutput};
use crate::pi_proxy::{PiProxyConfig, PiProxyError, PiProxyPlatformScope};
use crate::supervision::SandboxdSupervisorHandle;

const PI_RPC_RESPONSE_TIMEOUT: Duration = Duration::from_secs(30);

pub(super) struct PiProxyState {
    pub(super) config: PiProxyConfig,
    pub(super) child: Mutex<Option<PiRpcChild>>,
    pub(super) command_lock: Mutex<()>,
    pub(super) event_subscribers: Mutex<Vec<Sender<String>>>,
    pub(super) keepalive_manager: Arc<Mutex<KeepaliveManager>>,
    pub(super) active: AtomicBool,
    pub(super) activity_monitor_running: AtomicBool,
    pub(super) next_id: AtomicU64,
    pub(super) idempotency_store: Option<SharedIdempotencyStore>,
    pub(super) supervisor_handle: SandboxdSupervisorHandle,
    pub(super) platform_scope: Option<PiProxyPlatformScope>,
}

impl PiProxyState {
    fn next_pi_request_id(&self) -> String {
        let next = self.next_id.fetch_add(1, Ordering::Relaxed);
        format!("mistle_pi_{next}")
    }

    pub(super) fn send_pi_command(&self, command: Value) -> Result<Value, PiProxyError> {
        self.send_pi_command_with_events(command, None)
    }

    pub(super) fn send_pi_command_with_captured_events(
        &self,
        command: Value,
        captured_events: &mut Vec<Value>,
    ) -> Result<Value, PiProxyError> {
        self.send_pi_command_with_events(command, Some(captured_events))
    }

    fn send_pi_command_with_events(
        &self,
        mut command: Value,
        mut captured_events: Option<&mut Vec<Value>>,
    ) -> Result<Value, PiProxyError> {
        let _command_guard = self.command_lock.lock().map_err(|_| {
            PiProxyError::InvalidRequest("Pi command lock was poisoned".to_string())
        })?;
        let id = self.next_pi_request_id();
        command["id"] = Value::String(id.clone());
        let line = format!("{command}\n");
        let deadline = Instant::now() + PI_RPC_RESPONSE_TIMEOUT;
        {
            let mut guard = self.child.lock().map_err(|_| {
                PiProxyError::InvalidRequest("Pi child lock was poisoned".to_string())
            })?;
            let child = guard.as_mut().ok_or_else(|| {
                PiProxyError::InvalidRequest("Pi RPC process is not running".to_string())
            })?;
            if let Err(error) = child.stdin.write_all(line.as_bytes()) {
                self.mark_pi_rpc_process_restarting(error.to_string());
                return Err(PiProxyError::WritePi(error));
            }
            if let Err(error) = child.stdin.flush() {
                self.mark_pi_rpc_process_restarting(error.to_string());
                return Err(PiProxyError::WritePi(error));
            }
        }

        loop {
            let remaining = deadline.saturating_duration_since(Instant::now());
            if remaining.is_zero() {
                return Err(PiProxyError::PiResponseTimeout(id));
            }
            let output = {
                let guard = self.child.lock().map_err(|_| {
                    PiProxyError::InvalidRequest("Pi child lock was poisoned".to_string())
                })?;
                let child = guard.as_ref().ok_or_else(|| {
                    PiProxyError::InvalidRequest("Pi RPC process is not running".to_string())
                })?;
                child.receiver.recv_timeout(remaining)
            };
            match output {
                Ok(PiRpcOutput::Line(value)) => {
                    self.update_activity_from_pi_output(&value);
                    if value["type"] == "response" && value["id"] == id {
                        if value["success"] == true {
                            return Ok(value.get("data").cloned().unwrap_or(Value::Null));
                        }
                        let message = value["error"]
                            .as_str()
                            .unwrap_or("Pi RPC command failed")
                            .to_string();
                        return Err(PiProxyError::InvalidRequest(message));
                    }
                    if let Some(events) = captured_events.as_deref_mut() {
                        events.push(value);
                    } else {
                        self.broadcast_pi_event(value);
                    }
                }
                Ok(PiRpcOutput::Error(error)) => {
                    self.mark_pi_rpc_process_restarting(error.clone());
                    return Err(PiProxyError::InvalidRequest(error));
                }
                Ok(PiRpcOutput::Eof) => {
                    let message = "Pi RPC process stdout closed".to_string();
                    self.mark_pi_rpc_process_restarting(message.clone());
                    return Err(PiProxyError::InvalidRequest(message));
                }
                Err(RecvTimeoutError::Timeout) => return Err(PiProxyError::PiResponseTimeout(id)),
                Err(RecvTimeoutError::Disconnected) => {
                    let message = "Pi RPC reader disconnected".to_string();
                    self.mark_pi_rpc_process_restarting(message.clone());
                    return Err(PiProxyError::InvalidRequest(message));
                }
            }
        }
    }

    pub(super) fn subscribe_pi_events(&self) -> Receiver<String> {
        let (sender, receiver) = mpsc::channel();
        if let Ok(mut subscribers) = self.event_subscribers.lock() {
            subscribers.push(sender);
        }
        receiver
    }

    fn broadcast_pi_event(&self, event: Value) {
        let notification = render_pi_event_json_rpc_notification(event);
        if let Ok(mut subscribers) = self.event_subscribers.lock() {
            subscribers.retain(|sender| sender.send(notification.clone()).is_ok());
        }
    }

    pub(super) fn read_session_file(state_value: &Value) -> Result<&str, PiProxyError> {
        state_value["sessionFile"]
            .as_str()
            .ok_or(PiProxyError::MissingSessionFile)
    }

    pub(super) fn switch_session(
        &self,
        session_file: &str,
        captured_events: &mut Vec<Value>,
    ) -> Result<(), PiProxyError> {
        self.send_pi_command_with_captured_events(
            json!({ "type": "switch_session", "sessionPath": session_file }),
            captured_events,
        )?;
        Ok(())
    }

    fn update_activity_from_pi_output(&self, value: &Value) {
        if value["type"] == "agent_start" {
            self.set_active(true);
            return;
        }
        if value["type"] == "agent_end" {
            self.set_active(false);
            return;
        }
        if value["type"] == "response"
            && value["command"] == "get_state"
            && value["success"] == true
        {
            let state = &value["data"];
            let active = state["isStreaming"].as_bool().unwrap_or(false)
                || state["isCompacting"].as_bool().unwrap_or(false)
                || state["pendingMessageCount"].as_u64().unwrap_or(0) > 0;
            self.set_active(active);
        }
    }

    pub(super) fn set_active(&self, active: bool) {
        let previous = self.active.swap(active, Ordering::Relaxed);
        if previous == active {
            return;
        }
        if let Ok(mut keepalive_manager) = self.keepalive_manager.lock() {
            keepalive_manager.set_platform_active(active);
        }
    }

    pub(super) fn start_activity_monitor(state: Arc<Self>) -> bool {
        if state
            .activity_monitor_running
            .compare_exchange(false, true, Ordering::AcqRel, Ordering::Acquire)
            .is_err()
        {
            return false;
        }

        thread::spawn(move || {
            while state.active.load(Ordering::Relaxed) {
                if state
                    .send_pi_command(json!({ "type": "get_state" }))
                    .is_err()
                {
                    state.set_active(false);
                    break;
                }
                thread::sleep(Duration::from_secs(1));
            }
            state
                .activity_monitor_running
                .store(false, Ordering::Release);
            if state.active.load(Ordering::Relaxed) {
                Self::start_activity_monitor(state);
            }
        });
        true
    }

    pub(super) fn mark_active_and_start_activity_monitor(state: &Arc<Self>) {
        state.set_active(true);
        Self::start_activity_monitor(state.clone());
    }

    pub(super) fn register_pi_platform_root_pid(&self, pid: u32) -> Result<(), PiProxyError> {
        let Some(platform_scope) = &self.platform_scope else {
            return Ok(());
        };
        attach_pid_to_scope(&platform_scope.scope_paths, pid)
            .map_err(|error| PiProxyError::PlatformScope(error.to_string()))?;
        platform_scope
            .registry
            .replace_scope(
                platform_scope.registry_key.clone(),
                platform_scope.process_key.clone(),
                platform_scope.scope_paths.clone(),
                pid,
            )
            .map_err(PiProxyError::PlatformScope)
    }

    pub(super) fn kill_and_remove_pi_platform_scope(&self) -> Result<(), PiProxyError> {
        let Some(platform_scope) = &self.platform_scope else {
            return Ok(());
        };
        kill_scope(&platform_scope.scope_paths)
            .map_err(|error| PiProxyError::PlatformScope(error.to_string()))?;
        platform_scope
            .registry
            .remove_scope(&platform_scope.registry_key)
            .map_err(PiProxyError::PlatformScope)
    }
}

File: packages/sandboxd/piproxy/state.go
package piproxy

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"sync"
	"sync/atomic"
	"time"

	"github.com/mistle/sandboxd/cgroups"
	"github.com/mistle/sandboxd/keepalive"
	"github.com/mistle/sandboxd/supervision"
)

const piRPCResponseTimeout = 30 * time.Second

type State struct {
	config           Config
	child            *piRPCChild
	childMutex       sync.Mutex
	commandMutex     sync.Mutex
	eventSubscribers []chan []byte
	eventMutex       sync.Mutex
	keepaliveManager *keepalive.SharedManager
	idempotencyStore *SharedIdempotencyStore
	active           atomic.Bool
	activityMonitor  atomic.Bool
	nextID           atomic.Uint64
	supervisorHandle *supervision.SandboxdSupervisorHandle
	platformScope    *PlatformScope
}

type piRPCChild struct {
	command  *exec.Cmd
	stdin    io.WriteCloser
	output   chan piRPCOutput
	cwd      *string
	readerWG sync.WaitGroup
	done     chan error
}

type piRPCOutput struct {
	value map[string]any
	err   error
	eof   bool
}

func NewState(
	config Config,
	keepaliveManager *keepalive.SharedManager,
	supervisorHandle *supervision.SandboxdSupervisorHandle,
	idempotencyStore *SharedIdempotencyStore,
	platformScope *PlatformScope,
) *State {
	state := &State{
		config:           config,
		keepaliveManager: keepaliveManager,
		idempotencyStore: idempotencyStore,
		supervisorHandle: supervisorHandle,
		platformScope:    platformScope,
	}
	state.nextID.Store(1)
	return state
}

func (state *State) EnsureChild(cwd *string) error {
	state.childMutex.Lock()
	defer state.childMutex.Unlock()
	if state.child != nil {
		if !state.childExited(state.child) && !shouldReplaceChildForCWD(state.child, cwd) {
			return nil
		}
		if state.childExited(state.child) {
			state.markPiRPCProcessRestarting("Pi RPC process exited")
		}
	}
	if state.child != nil {
		child := state.child
		state.child = nil
		state.terminateChild(child)
		state.setActive(false)
		state.markPiRPCProcessStopped()
		if err := state.killAndRemovePlatformScope(); err != nil {
			return err
		}
	}
	state.markPiRPCProcessStarting()
	command := exec.Command(state.config.PiCLIPath, "--mode", "rpc")
	if cwd != nil {
		command.Dir = *cwd
	}
	command.Env = commandEnvironment(state.config.Env)
	command.Stderr = os.Stderr
	stdin, err := command.StdinPipe()
	if err != nil {
		state.markPiRPCProcessRestarting(err.Error())
		return fmt.Errorf("spawned Pi RPC process did not expose stdin: %w", err)
	}
	stdout, err := command.StdoutPipe()
	if err != nil {
		state.markPiRPCProcessRestarting(err.Error())
		return fmt.Errorf("spawned Pi RPC process did not expose stdout: %w", err)
	}
	if err := command.Start(); err != nil {
		state.markPiRPCProcessRestarting(err.Error())
		_ = state.killAndRemovePlatformScope()
		return fmt.Errorf("failed to spawn Pi RPC process: %w", err)
	}
	if err := state.registerPlatformScope(uint32(command.Process.Pid)); err != nil {
		state.markPiRPCProcessRestarting(err.Error())
		_ = command.Process.Kill()
		_ = state.killAndRemovePlatformScope()
		return err
	}
	child := &piRPCChild{
		command: command,
		stdin:   stdin,
		output:  make(chan piRPCOutput, 64),
		cwd:     cloneStringPointer(cwd),
		done:    make(chan error, 1),
	}
	child.readerWG.Add(1)
	go readPiRPCStdout(stdout, child.output, &child.readerWG)
	go func() {
		child.done <- command.Wait()
		close(child.done)
	}()
	state.child = child
	state.markPiRPCProcessHealthy(command.Process.Pid)
	return nil
}

func (state *State) ShutdownChild() error {
	state.childMutex.Lock()
	child := state.child
	state.child = nil
	state.childMutex.Unlock()
	if child == nil {
		return nil
	}
	state.terminateChild(child)
	state.setActive(false)
	state.markPiRPCProcessStopped()
	return state.killAndRemovePlatformScope()
}

func (state *State) SendCommand(command map[string]any) (map[string]any, error) {
	return state.SendCommandWithCapturedEvents(command, nil)
}

func (state *State) SendCommandWithCapturedEvents(command map[string]any, capturedEvents *[]map[string]any) (map[string]any, error) {
	state.commandMutex.Lock()
	defer state.commandMutex.Unlock()
	id := state.nextPiRequestID()
	command["id"] = id
	serialized, err := json.Marshal(command)
	if err != nil {
		return nil, err
	}
	serialized = append(serialized, '\n')
	child, err := state.currentChild()
	if err != nil {
		return nil, err
	}
	if _, err := child.stdin.Write(serialized); err != nil {
		state.markPiRPCProcessRestarting(err.Error())
		return nil, fmt.Errorf("failed to write Pi RPC command: %w", err)
	}

	deadline := time.After(piRPCResponseTimeout)
	for {
		select {
		case output := <-child.output:
			if output.err != nil {
				state.markPiRPCProcessRestarting(output.err.Error())
				return nil, output.err
			}
			if output.eof {
				message := "Pi RPC process stdout closed"
				state.markPiRPCProcessRestarting(message)
				return nil, errors.New(message)
			}
			state.updateActivityFromPiOutput(output.value)
			if output.value["type"] == "response" && output.value["id"] == id {
				if output.value["success"] == true {
					if data, ok := output.value["data"].(map[string]any); ok {
						return data, nil
					}
					return map[string]any{}, nil
				}
				if message, ok := output.value["error"].(string); ok && message != "" {
					return nil, errors.New(message)
				}
				return nil, fmt.Errorf("Pi RPC command failed")
			}
			if capturedEvents != nil {
				*capturedEvents = append(*capturedEvents, output.value)
			} else {
				state.BroadcastEvent(output.value)
			}
		case <-deadline:
			return nil, fmt.Errorf("timed out waiting for Pi RPC response %q", id)
		}
	}
}

func (state *State) SubscribeEvents() <-chan []byte {
	channel := make(chan []byte, 64)
	state.eventMutex.Lock()
	state.eventSubscribers = append(state.eventSubscribers, channel)
	state.eventMutex.Unlock()
	return channel
}

func (state *State) BroadcastEvent(event map[string]any) {
	notification := RenderPiEventJSONRPCNotification(event)
	state.eventMutex.Lock()
	defer state.eventMutex.Unlock()
	subscribers := state.eventSubscribers[:0]
	for _, subscriber := range state.eventSubscribers {
		select {
		case subscriber <- notification:
			subscribers = append(subscribers, subscriber)
		default:
			subscribers = append(subscribers, subscriber)
		}
	}
	state.eventSubscribers = subscribers
}

func (state *State) SwitchSession(sessionFile string, capturedEvents *[]map[string]any) error {
	_, err := state.SendCommandWithCapturedEvents(map[string]any{
		"type":        "switch_session",
		"sessionPath": sessionFile,
	}, capturedEvents)
	return err
}

func (state *State) MarkActiveAndStartActivityMonitor() {
	state.setActive(true)
	if !state.activityMonitor.CompareAndSwap(false, true) {
		return
	}
	go func() {
		defer state.activityMonitor.Store(false)
		for state.active.Load() {
			if _, err := state.SendCommand(map[string]any{"type": "get_state"}); err != nil {
				state.setActive(false)
				return
			}
			time.Sleep(time.Second)
		}
	}()
}

func (state *State) StartChildMonitor(stop <-chan struct{}) {
	ticker := time.NewTicker(250 * time.Millisecond)
	defer ticker.Stop()
	for {
		select {
		case <-stop:
			return
		case <-ticker.C:
			state.childMutex.Lock()
			child := state.child
			exited := child != nil && state.childExited(child)
			state.childMutex.Unlock()
			if exited {
				_ = state.EnsureChild(nil)
			}
		}
	}
}

func (state *State) currentChild() (*piRPCChild, error) {
	state.childMutex.Lock()
	defer state.childMutex.Unlock()
	if state.child == nil {
		return nil, fmt.Errorf("Pi RPC process is not running")
	}
	return state.child, nil
}

func (state *State) nextPiRequestID() string {
	next := state.nextID.Add(1) - 1
	return fmt.Sprintf("mistle_pi_%d", next)
}

func (state *State) terminateChild(child *piRPCChild) {
	terminateCommand(child.command, child.done)
	child.readerWG.Wait()
}

func (state *State) childExited(child *piRPCChild) bool {
	select {
	case _, ok := <-child.done:
		if !ok {
			return true
		}
		return true
	default:
		return false
	}
}

func (state *State) registerPlatformScope(pid uint32) error {
	if state.platformScope == nil {
		return nil
	}
	if err := cgroups.AttachPIDToScope(state.platformScope.ScopePaths, pid); err != nil {
		return fmt.Errorf("failed to manage Pi platform scope: %w", err)
	}
	if err := state.platformScope.Registry.ReplaceScope(
		state.platformScope.RegistryKey,
		state.platformScope.ProcessKey,
		state.platformScope.ScopePaths,
		pid,
	); err != nil {
		return fmt.Errorf("failed to manage Pi platform scope: %w", err)
	}
	return nil
}

func (state *State) killAndRemovePlatformScope() error {
	if state.platformScope == nil {
		return nil
	}
	if err := cgroups.KillScope(state.platformScope.ScopePaths); err != nil {
		return fmt.Errorf("failed to manage Pi platform scope: %w", err)
	}
	if err := state.platformScope.Registry.RemoveScope(state.platformScope.RegistryKey); err != nil {
		return fmt.Errorf("failed to manage Pi platform scope: %w", err)
	}
	return nil
}

func (state *State) setActive(active bool) {
	previous := state.active.Swap(active)
	if previous == active {
		return
	}
	state.keepaliveManager.WithLocked(func(manager *keepalive.Manager) {
		manager.SetPlatformActive(active)
	})
}

func (state *State) updateActivityFromPiOutput(value map[string]any) {
	switch value["type"] {
	case "agent_start":
		state.setActive(true)
	case "agent_end":
		state.setActive(false)
	case "response":
		if value["command"] == "get_state" && value["success"] == true {
			data, _ := value["data"].(map[string]any)
			active := boolValue(data["isStreaming"]) ||
				boolValue(data["isCompacting"]) ||
				numberValue(data["pendingMessageCount"]) > 0
			state.setActive(active)
		}
	}
}

func (state *State) markPiRPCProcessStarting() {
	if !state.supervisorHandle.TracksComponent(supervision.ComponentPiRpcProcess) {
		return
	}
	state.supervisorHandle.ReplaceComponentDetails(supervision.ComponentPiRpcProcess, map[string]string{
		"cliPath": state.config.PiCLIPath,
	})
	state.supervisorHandle.MarkComponentStarting(supervision.ComponentPiRpcProcess)
}

func (state *State) markPiRPCProcessHealthy(pid int) {
	if !state.supervisorHandle.TracksComponent(supervision.ComponentPiRpcProcess) {
		return
	}
	state.supervisorHandle.ReplaceComponentDetails(supervision.ComponentPiRpcProcess, map[string]string{
		"cliPath": state.config.PiCLIPath,
		"pid":     fmt.Sprint(pid),
	})
	state.supervisorHandle.MarkComponentHealthy(supervision.ComponentPiRpcProcess)
}

func (state *State) markPiRPCProcessRestarting(errorMessage string) {
	if !state.supervisorHandle.TracksComponent(supervision.ComponentPiRpcProcess) {
		return
	}
	state.supervisorHandle.MarkComponentRestarting(supervision.ComponentPiRpcProcess, errorMessage)
}

func (state *State) markPiRPCProcessStopped() {
	if !state.supervisorHandle.TracksComponent(supervision.ComponentPiRpcProcess) {
		return
	}
	state.supervisorHandle.MarkComponentStopped(supervision.ComponentPiRpcProcess)
}

func readPiRPCStdout(stdout io.Reader, output chan<- piRPCOutput, waitGroup *sync.WaitGroup) {
	defer waitGroup.Done()
	scanner := bufio.NewScanner(stdout)
	for scanner.Scan() {
		var value map[string]any
		if err := json.Unmarshal(scanner.Bytes(), &value); err != nil {
			output <- piRPCOutput{err: err}
			continue
		}
		output <- piRPCOutput{value: value}
	}
	if err := scanner.Err(); err != nil {
		output <- piRPCOutput{err: err}
		return
	}
	output <- piRPCOutput{eof: true}
}

func shouldReplaceChildForCWD(child *piRPCChild, requestedCWD *string) bool {
	return child.cwd == nil && requestedCWD != nil
}

func cloneStringPointer(value *string) *string {
	if value == nil {
		return nil
	}
	clone := *value
	return &clone
}

func boolValue(value any) bool {
	result, _ := value.(bool)
	return result
}

func numberValue(value any) float64 {
	result, _ := value.(float64)
	return result
}
